### PR TITLE
feat: Add kosmos pod controller interface

### PR DESCRIPTION
**Pull request by OrangeBao**

### CODE REVIEW

File: pkg/clusterrouter/kresourcemanager/knodemanager.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ import (
 
 	"github.com/kosmos.io/kosmos/cmd/knode-manager/app/config"
 	knodev1alpha1 "github.com/kosmos.io/kosmos/pkg/apis/kosmos/v1alpha1"
-	knode2 "github.com/kosmos.io/kosmos/pkg/clusterrouter/knodemanager/knode"
+	knode2 "github.com/kosmos.io/kosmos/pkg/clusterrouter/kresourcemanager/knode"
```
Comment (wuyingjun-lucky) [inline pkg/clusterrouter/kresourcemanager/knodemanager.go]: knode2? seems not good

File: pkg/clusterrouter/kresourcemanager/kpod/sync.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+// This is used to clean up keys for deleted pods after they have been fully deleted in the API server.
+type syncWrapper interface {
+	_deletePodKey(context.Context, string)
+}
+
+func (p *syncProviderWrapper) NotifyPods(ctx context.Context, f func(*corev1.Pod)) {
+	p.notify = f
+}
+
+func (p *syncProviderWrapper) _deletePodKey(ctx context.Context, key string) {
```
Comment (duanmengkk) [inline pkg/clusterrouter/kresourcemanager/kpod/sync.go]: why the func start with prefix "_"

File: pkg/clusterrouter/plugins/virtualk8s/node.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog"
+
+	"github.com/kosmos.io/kosmos/pkg/clusterrouter/common"
+	"github.com/kosmos.io/kosmos/pkg/clusterrouter/utils"
+)
+
+// ConfigureNode enables a provider to configure the node object that
+// will be used for Kubernetes.
+func (v *VirtualK8S) ConfigureNode(ctx context.Context, node *corev1.Node) {
```
Comment (duanmengkk) [inline pkg/clusterrouter/plugins/virtualk8s/node.go]: VirtualK8S can rename to anther name

File: pkg/clusterrouter/utils/log/log.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	L Logger = nopLogger{}
+)
+
+type (
+	loggerKey struct{}
+)
+
+// Logger is the interface used for logging in virtual-kubelet
+//
+// virtual-kubelet will access the logger via context using `GetLogger` (or its alias, `G`)
```
Comment (duanmengkk) [inline pkg/clusterrouter/utils/log/log.go:43]: virtual-kubelet

File: pkg/clusterrouter/utils/log/logrus/logrus.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package logrus implements a github.com/virtual-kubelet/virtual-kubelet/log.Logger using Logrus as a backend
+// You can use this by creating a logrus logger and calling `FromLogrus(entry)`.
+// If you want this to be the default logger for virtual-kubelet, set `log.L` to the value returned by `FromLogrus`
```
Comment (duanmengkk) [inline pkg/clusterrouter/utils/log/logrus/logrus.go]: some comment should be changed

File: pkg/clusterrouter/utils/trace/trace.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	ctx, span := tracer.StartSpan(ctx, name)
+	ctx = log.WithLogger(ctx, span.Logger())
+	return ctx, span
+}
+
+// Span encapsulates a tracing event
+type Span interface {
+	End()
```
Comment (duanmengkk) [inline pkg/clusterrouter/utils/trace/trace.go:66]: Span is too simalar like vk

File: pkg/clusterrouter/kresourcemanager/knodemanager.go (hunk 1 above)
Comment (OrangeBao) [reply pkg/clusterrouter/kresourcemanager/knodemanager.go:21]: I didn't change this word

File: pkg/clusterrouter/kresourcemanager/kpod/sync.go (hunk 2 above)
Comment (OrangeBao) [reply pkg/clusterrouter/kresourcemanager/kpod/sync.go]: This part of the code was copied from another project

File: pkg/clusterrouter/plugins/virtualk8s/node.go (hunk 3 above)
Comment (OrangeBao) [reply pkg/clusterrouter/plugins/virtualk8s/node.go]: ok

File: pkg/clusterrouter/utils/trace/trace.go (hunk 6 above)
Comment (OrangeBao) [reply pkg/clusterrouter/utils/trace/trace.go]: This part has been deleted

File: pkg/clusterrouter/utils/log/log.go (hunk 4 above)
Comment (OrangeBao) [reply pkg/clusterrouter/utils/log/log.go]: All comments have been deleted

File: pkg/clusterrouter/kresourcemanager/knodemanager.go (hunk 1 above)
Comment (OrangeBao) [reply pkg/clusterrouter/kresourcemanager/knodemanager.go]: ok

File: pkg/clusterrouter/kresourcemanager/kpod/sync.go (hunk 2 above)
Comment (OrangeBao) [reply pkg/clusterrouter/kresourcemanager/kpod/sync.go]: I get  it

File: pkg/clusterrouter/utils/log/logrus/logrus.go (hunk 5 above)
Comment (OrangeBao) [reply pkg/clusterrouter/utils/log/logrus/logrus.go]: I get  it

File: pkg/knode-manager/adapters/k8s/node.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+)
+
+type NodeAdapter struct {
+}
+
+func NewNodeAdapter() (*NodeAdapter, error) {
+	return nil, network.ErrNotImplemented
+}
+
+func (v *NodeAdapter) Configure(ctx context.Context, node *corev1.Node) {
```
Comment (xyz2277) [inline pkg/knode-manager/adapters/k8s/node.go]: n is better

File: pkg/knode-manager/adapters/k8s/pod.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+type PodAdapter struct {
+}
+
+func NewPodAdapter() (*PodAdapter, error) {
+	return nil, network.ErrNotImplemented
+}
+
+func (v *PodAdapter) Create(ctx context.Context, pod *corev1.Pod) error {
```
Comment (xyz2277) [inline pkg/knode-manager/adapters/k8s/pod.go]: p is better

**Comment by wuyingjun-lucky** [comment]
/retitle feat: Add kosmos pod controller interface

**Comment by OrangeBao** [comment]
> /retitle feat: Add kosmos pod controller interface

ok

**Pull request merged**